### ISSUE TITLE
Native support for coprocesses/daemons/long-running helpers

### ISSUE THREAD

**Issue by Screwtapello** [mawww/kakoune#4127]
### Feature 

Kakoune should natively support managing coprocesses, long-running external processes that communicate bi-directionally with Kakoune. This would involve:

- A way to tell Kakoune to launch a shell as a coprocess, holding onto its stdin, stdout, and stderr.
- A way to send a string to the coprocess' stdin
- Routing the coprocess' stderr to `*debug*` like any other external process
- A way for the coprocess to send commands back to Kakoune (such as writing to stdout)
- A way to kill a misbehaving/stuck coprocess

### Rationale

Kakoune's `%sh{}` expansions provide excellent support for interacting with short-running external processes. It's more difficult, but still reasonable, to interact with long-running one-shot processes (for example, the way [grep.kak](https://github.com/mawww/kakoune/blob/212242cc602d386fe1b7c1ce10c55ad0430c4960/rc/tools/grep.kak#L17) launches a grep process in the background, plumbed to a FIFO buffer, or [lint.kak](https://github.com/mawww/kakoune/blob/212242cc602d386fe1b7c1ce10c55ad0430c4960/rc/tools/lint.kak) launches a subshell in the background that pipes to `kak -p` when its work is done).

However, every plugin that wants to interact with a long-running external process (a "coprocess") has to solve the same problems:

- launching the external process, and shutting it down when the Kakoune session ends
- de-multiplexing individual messages from Kakoune into per-session event streams
- communicating events back to Kakoune
- finding a way to report errors to the user

Not only is this effort duplicated in every plugin that wants a coprocess, but it's also a substantial increase in complexity over just running a process in the background:

- launching a process from Kakoune is easy, but shutting it down requires a timeout (short timeouts make the coprocess lose state easily, long timeouts risk having a heavy coprocess hang around long after the session exits)
- De-multiplexing messages requires using a FIFO for rendezvous (tricky, since FIFOs are supposed to be one-shot) or a socket (you have to write a client as well as a server, since POSIX provides no command-line socket-connection tool)
- Communicating events back to Kakoune is straightforward enough with `kak -p`, but expensive
- You can write errors to a log-file easily enough, but who's going to rotate it? How do you keep errors for different sessions separate? What if the error is "cannot open log file for writing"?

If these things were provided by Kakoune, it would make writing coprocess-based plugins much, much easier.

### Concrete use-cases

- `kak-lsp` goes to a *lot* of effort to ensure it gets launched at the right times, and shuts down eventually. It uses a socket to rendezvous, and the `kak-lsp` binary contains both "client" and "server" functionality. Historically it's had issues with leaving a crazy number of kak-lsp processes running in the background when something goes wrong.
- `kakoune-repl-buffer` uses a special helper-script with some low-level POSIX trickery to make FIFO rendezvous work.It depends on the FIFO being cleaned up by a hook, which may not run if Kakoune does not exit cleanly.
- The other day somebody dropped by the IRC channel asking for help with a plugin that would capture mouse press and release events, and generate "click", "double-click", and "triple-click" events from them. The state tracking seemed like a fun Sunday afternoon activity, until I remembered all the boilerplate process-management that would have to go along with it.

### Strawman design

A new kind of option, `coprocess`. The user-settable value is of the form `<pid>|<command>`, where `<pid>` should always be empty when setting the option, and `<command>` is a shell command to launch the coprocess (including the usual Kakoune shell variables like `$kak_session`, etc.).

A new command, `coprocess-send-text <coprocess> <text>`, where `<coprocess>` is the name of a coprocess-type option, and `<text>` is arbitrary text to send. If the coprocess is not yet running, Kakoune launches the command, holds onto its stdin/stdout/stderr file descriptors, and fills in the PID in the option's value. Once the coprocess is running, `<text>` is written to the coprocess' stdin.

Kakoune will listen to the coprocess' stderr descriptior, and any time any data becomes available, it gets added to the `*debug*` buffer, wrapped in `coprocess OPTIONNAME stderr: <<< >>>`. `OPTIONNAME` should be the name of the coprocess option.

Kakoune will listen to the coprocess' stdout descriptor, and any time data becomes available, it is parsed as Kakoune commands (as if it were sent to `kak -p`).

If Kakoune is notified that a coprocess has exited (for example, by SIGCHLD or by polling `waitpid()` with `WNOHANG`), it should remove the `<pid>` field from the associated option.

Whenever a coprocess option's value goes out of scope (for example, it was set at buffer scope and the buffer was deleted, or a new value was set at the same scope as the current value), Kakoune should shut the coprocess' stdin, if it was running. If a new value is set in a child scope while a coprocess is running at an outer scope, the outer scope's coprocess should not be killed.

If a coprocess needs to be killed manually, it's easy enough to write a command like:

```
def lsp-kill %{
    eval %sh{
        case "$kak_opt_kak_lsp_coproc" in
            |*)
                echo "echo kak-lsp not running"
                ;;
            *|*)
                kill -TERM ${kak_opt_kak_lsp_coproc%%|*}
                ;;
        esac
    }
}
```

**Comment by danr** [comment]
Such an addition would be very helpful. Here are some notes:

One issue with communicating via stdio is separating individual messages.
Headers could be used: should sending to a process also add a header `Content-length: <bytes>`.
Null-byte separation could be used: `<message>\0<message>`.
Newline separation is another option which is probably most "natural", then coprocess-send-text could quote the newline. 

The introduced command `coprocess-send-text <coprocess> <text>` could be a new flag `-to-process` to the `echo` command: `echo -to-process <coprocess> <text>`

The strawman design only supports asynchronous messages. I think sync calls are sometimes to be preferred,
it ensures that the reply is coherent with the request and that the client doesn't make too many requests at once.

**Comment by Guest0x0** [comment]
Would love to see such an addition.

From my experience on writing some interactive plugins, I find the following things particularly painful:
- FIFOs are *really fun* to work with
- executing a command that is sensible only under window context, such as moving the viewport or modifying the selection,
requires additional book-keeping to maintain a list of clients for some specific buffer.



> One issue with communicating via stdio is separating individual messages

I think this issue can be solved by wrapping the actual process in a simple relay script, for example:
```
set-option buffer myproc %{
    the-process | sed -n 'reformat-command'
}
```
supporting full shell syntax in command may require some special attention in terms of option formatting for coprocess, though.


> The strawman design only supports asynchronous messages. I think sync calls are sometimes to be preferred,
it ensures that the reply is coherent with the request and that the client doesn't make too many requests at once.

I agree with that. Perhaps we can add a `-wait` option to `coprocess-send-text`, which forces kakoune to wait for the next output from the coprocess, with an optional timeout in case the process hangs up?


BTW, introducing new hooks for coprocess launch and exit seems desirable, too.

**Comment by danr** [comment]
Having given this some thought, here is my tinman proposal:

The command to define coprocesses is `define-coprocess <name> <shell-block>`. Kak executes the shell block in the background and gives it an environment variable `"$kak_coprocess_id"`. The `kak` executable now has two new flags. The coprocess will register itself as ready using `kak --coprocess-ready "$kak_coprocess_id"` which makes the coprocess block untill it is called. If it is called then kak blocks until the coprocess executes `kak --coprocess-reply "$kak_coprocess_id"`, executing kak commands on stdin. This is to support synchronous replies.

This is an example of a counting coprocess:

```sh
define-coprocess counting-example %{
    # this will be executed by the shell

    # some initialization!
    i=0

    while true; do
        # this blocks until the coprocess is called:
        kak -s "$kak_session" --coprocess-ready "$kak_coprocess_id"

        # now kak is blocked, just like how %sh-blocks block kak,
        # and we can prepare our response:
        i=$(expr "$i" + 1)

        # we now pipe to kak saying that this round of the coprocess is ready
        printf %s "info 'calls: $i'" | kak -s "$kak_session" --coprocess-reply "$kak_coprocess_id"
        # kak executes this and is unblocked again

        # (the coprocess could communicate asynchronously with kak here using `kak -p` as normal)
    done
}
```

We can now call the coprocess using call-coprocess:

```sh
call-coprocess counting-example # info box with `calls: 1`!
```

and again:

```sh
call-coprocess counting-example # info box with `calls: 2`!
```

Of course we want to be able to pass arguments to the coprocess as well. We add a flag `--coprocess-context` to the `kak --coprocess-ready` command which forwards a kak-interpolated string, such as `%arg{@}`. This is prepared just like a call to `echo -quoting shell`.

With this at hand let's extend the example to count the current selection, and also support reset:

```sh
define-coprocess -params 0..1 count-selection-example %{
    # this will be executed by the shell

    # some initialization!
    i=0

    while true; do
        # this blocks until the coprocess is called:
        eval set -- "$(
            kak -s "$kak_session" --coprocess-ready "$kak_coprocess_id" --coprocess-context '%arg{1} %val{selection}'
        )"
        # now:
        #   $1: %arg{1}
        #   $2: %val{selection}

        if test "$1" = "reset"; then
            i=0
        else
            i=$(expr "$i" + "$2")
        fi

        printf %s "info 'count: $i'" | kak -s "$kak_session" --coprocess-reply "$kak_coprocess_id"
    done
}
```

So say we have the selection at the string `5` and call:

```sh
call-coprocess count-selection-example # info box with `count: 5`!
```

but we can also do:

```sh
call-coprocess count-selection-example reset # info box with `count: 0`!
```

Kak of course keeps track of if the coprocess pid and checks if it is dead and restarts it if so.  And just like the user waits for %sh-blocks to finish with an option to cancel it with `<c-c>` the user can cancel a call to a coprocess.  This could either when waiting for the coprocess to become ready (using `--coprocess-ready`), or at its reply (using `--coprocess-reply`).

**Comment by TeddyDD** [comment]
Relevant - implementation of this idea in user space https://github.com/tomKPZ/pykak

Still - probably should be in core.

**Comment by skewballfox** [comment]
What's the status of this? 

Has anything been added in the last two years that someone should make use of if they wanted to implement this?
